src/matcher/index.ts

```diff
@@ -71,11 +72,37 @@ export function createRouterMatcher(
     return matcherMap.get(name)
   }
 
+  function routeComponentsOptionGuard(
+    recordComponents?: Record<string, RawRouteComponent>
+  ): never | void {
+    // Guard for record.components. See Issue #1260
+    // Recieved components: Home
+    // This should be components: { default: Home, others } or component: Home
+    for (const name in recordComponents) {
+      const rawComponent = recordComponents[name]
+      // property components should be object
+      if (
+        !rawComponent ||
+        (typeof rawComponent !== 'object' && typeof rawComponent !== 'function')
+      ) {
+        warn(
+          `Components property in record is not valid. ` +
+            `Property "components" should be component: MyView or ` +
+            `components: { default: MyView }`
+        )
+        // same with navigation guards
+        throw new Error(`Invalid route record "components" property.`)
+      }
+    }
+  }
+
   function addRoute(
     record: RouteRecordRaw,
     parent?: RouteRecordMatcher,
     originalRecord?: RouteRecordMatcher
   ) {
+    // for route records.components guard
+    if (__DEV__) routeComponentsOptionGuard(record.components)
```

Comment:
then we could here change the `record.components` to the returned value
```suggestion
    if (__DEV__) {
      record.components = routeComponentsOptionGuard(record.components)
    }
```